docs/polymorphic-table-types.md

```diff
@@ -0,0 +1,198 @@
+# Polymorphic Table Types
+
+## Summary
+
+We propose implementing polymorphic tables of the form `<T>{ [F<T>] : G<T> }` analogous to already-existing polymorphic functions `<T>(F<T>) -> G<T>`.
+
+## Motivation
+
+The goal of this RFC is furthering type-safety and type-representation, especially in niche or performance-critical applications. Polymorphic Functions exist already in the form `<T>(F<T>) -> G<T>`, where `T` is inferred at the call site, with similar applications.
+
+At a type level, luau functions and tables are homomorphic as type mappings. One can view a table as a function mapping its keys to its values, or a function as a table with inputs as keys and outputs as values. This change resolves the hole left behind implementing polymorphic functions while excluding polymorphic tables, forcing type-safe applications to prefer functions over tables as maps.
+
+### General Indexer Pattern
+
+At an application level, the current workaround for this is to wrap the table in a function getter like so:
+
+```lua
+-- private
+local foo = {} :: { [F<any>]: G<any> }
+
+-- public
+local function getFoo<T>(k: F<T>): G<T>
+	return foo[k]
+end
+
+local function setFoo<T>(k: F<T>, v: G<T>)
+	foo[k] = v
+end
+
+local iterFoo = pairs(foo)
+```
+
+This has five immediate consequences.
+
+1. This pattern extends to every table of this nature, increasing surface area, duplication, and maintenance.
+2. In the event the functions don't get inlined, we incur an extra performance penalty per read and write, disallowing certain applications type-safety.
+
+If we are using a function then one might try and take advantage of the extra indirection by hiding the table implementation to prevent misuse. This raises the following consequences:
+
+3. The implementation is opaque and can be modified later without worry.
+4. Mutations require yet another function, a setter, with the cons listed before.
+5. Iterations requires yet another function, or at least stored iterator, reinforcing the cons listed before.
+
+In short, tables are mutable and iterable, functions are not.
```

Comment:
These two features of tables make polymorphism _much_ more complex here. Mutability forces you to use invariance for table keys and values, and iteration has its own questions.